### ISSUE TITLE
Bind ACL includes to ActiveRecord::Query includes

### ISSUE THREAD

**Issue by calebfaruki** [waratuman/standardapi#53]
https://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-includes

Would be a huge performance improvement. Traverse `ActiveRecord::Reflection::AbstractReflection` inclusions and append it to controller queries.

https://github.com/waratuman/standardapi/blob/master/lib/standard_api/controller.rb#L33

Going for something like this...

```rb
association_includes = includes...
preloadables(
  resources.includes(association_includes).limit(limit).offset(params[:offset]).sort(orders),
  includes
)
```

**Comment by malomalo** [comment]
Are you seeing better performance with `.include` as opposed to `.preload`.

The `preloadables` function does `.preload`:

https://github.com/waratuman/standardapi/blob/91855bbf78167cb09565f8fd63b310666be30de8/lib/standard_api/helpers.rb#L22

**Comment by calebfaruki** [comment]
In general, the consensus is that `includes` will perform better than `preload` because `preload` will perform additional SQL queries by default. Whereas `includes` will do a `FULL OUTER JOIN` on inclusions in most cases. 

It's hard to pinpoint where in Rails it actually does this. But in general, if you know you need the associated records, it's best to do an `includes` unless it causes unreasonably expensive SQL query times.

Overall, `includes` seems like a modest compromise between `preload` and `eager_load`. Most articles I've read suggest not using `preload` unless doing a `LEFT OUTER JOIN` is prohibitively expensive.

Two articles discussing tradeoffs of `preload`, `includes`, and `eager_load`.
- https://scoutapm.com/blog/activerecord-includes-vs-joins-vs-preload-vs-eager_load-when-and-where
- https://tadhao.medium.com/joins-vs-preload-vs-includes-vs-eager-load-in-rails-5f721c44b3a9

**Comment by malomalo** [comment]
Ah yes, now I remember. Back in the day traversing the join tracker that keeps track of all this was difficult so was hard to use with `eager_load` to update there `where` clause, but it worked most of the time; not all.

That may have changed, I think the `activerecord-filter` can now work uses the tracker. We can try a branch with `eager_load` and run the test and see if it works, then potentially move to `includes` as the default.

At the same time `eager_load` can greatly increase the data response from the database, `eager_loading` has many will return the base model 100x if model has 100 children, so maybe we make it a setting?

So in summary:
[] Make branch with `eager_load` instead of `preload`
[] Make test pass and change to `includes`
[] Optionally make a setting to include via `includes`, `eager_load`, or `preload`

**Comment by calebfaruki** [comment]
Here's a good article: https://engineering.gusto.com/a-visual-guide-to-using-includes-in-rails/

Found it in this PR: https://github.com/rails/rails/pull/41438

The n+1 issue with `includes` vs `eager_load` doesn't really come up until you're querying an association of an association:


> :includes will default to use :preload unless you reference the association being loaded in a subsequent clause, such as :where or :order. When constructing a query this way, you also need to explicitly reference the eager loaded model.

As an example:

```
&include[parent][children][descendants]=true
```

In this case, you would probably want to use `eager_load` to optimize this query. You wouldn't always want to do this. And when you do it's probably because your use case necessitates it.

But then you could do a has_many/through to provide a hint to the ORM.

```rb
class Parent < ApplicationRecord
  has_many :descendants, through: :children
end
```

But if you have an association that can't use an ActiveRecord reference, you could probably affect the eager_load behavior here without relying on `standardapi` magic to do so, right?

```rb
class Parent < ApplicationRecord
  def children
    Descendant.filter(...)
  end
end
```

Rather than overengineer standardapi, would just be good to have some recipes!

**Comment by bemky** [comment]
We could allow the API consumer to decide which method to use. 

```
resource?preload[parent][child]=true
```

```
resource?eager_load[parent][child]=true
```

```
resource?include[parent][child]=true
```

Could support designating nested include method also:
```
resource?include[parent][eager_load][child]=true
```
